recipe-client-addon/lib/NormandyApi.jsm

```diff
@@ -145,20 +153,67 @@ this.NormandyApi = {
 
   /**
    * Fetch an array of available actions from the server.
+   * @param filters
+   * @param filters.enabled {boolean} If true, only returns enabled
+   * recipes. Default true.
    * @resolves {Array}
    */
-  async fetchActions() {
-    const actionApiUrl = await this.getApiUrl("action-list");
-    const res = await this.get(actionApiUrl);
-    return res.json();
+  async fetchRecipes(filters = {enabled: true}) {
+    return this.fetchSignedObjects("recipe", filters);
+  },
+
+  /**
+   * Fetch an array of available actions from the server.
+   * @resolves {Array}
+   */
+  async fetchActions(filters = {}) {
+    return this.fetchSignedObjects("action", filters);
   },
 
   async fetchImplementation(action) {
-    const response = await fetch(action.implementation_url);
-    if (response.ok) {
-      return response.text();
+    const implementationUrl = new URL(this.absolutify(action.implementation_url));
+
+    // fetch implementation
+    const response = await fetch(implementationUrl);
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch action implementation for ${action.name}: ${response.status}`
+      );
+    }
+    const responseText = await response.text();
+
+    // Try to verify integrity of the implementation text.  If the
+    // integrity value doesn't match the content or uses an unknown
+    // algorithm, fail.
+
+    // Get the last non-empty portion of the url path, and split it
+    // into two to get the aglorithm and hash.
+    const parts = implementationUrl.pathname.split("/");
+    const lastNonEmpty = parts.filter(p => p !== "").slice(-1)[0];
+    const [algorithm, ...hashParts] = lastNonEmpty.split("-");
+    const expectedHash = hashParts.join("-");
+
+    if (algorithm !== "sha384") {
```

Comment:
If you're only implement `sha384`, why store the name of the alrgorithm at all?
You could also take then `len()` of the hash to determine which algorithm was used. SHA384 hashes are 96 bytes long.